src/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/OidCollection.cs

```diff
@@ -3,26 +3,42 @@
 // See the LICENSE file in the project root for more information.
 
 using System.Collections;
-using System.Collections.Generic;
 using Internal.Cryptography;
 
 namespace System.Security.Cryptography
 {
     public sealed class OidCollection : ICollection
     {
-        public OidCollection()
-        {
-            _list = new List<Oid>();
-        }
+        private Oid[] _oids = Array.Empty<Oid>();
+        private int _count;
+
+        public OidCollection() { }
 
         public int Add(Oid oid)
         {
-            int count = _list.Count;
-            _list.Add(oid);
+            int count = _count;
+            if (count == _oids.Length)
+            {
+                Array.Resize(ref _oids, count == 0 ? 4 : count * 2);
+            }
+            _oids[count] = oid;
+            _count = count + 1;
             return count;
         }
 
-        public Oid this[int index] => _list[index];
+        public Oid this[int index]
+        {
+            get
+            {
+                if ((uint)index >= (uint)_count)
+                {
+                    // For compat, throw an ArgumentOutOfRangeException instead of
+                    // the IndexOutOfRangeException that comes from the array's indexer.
+                    throw new ArgumentOutOfRangeException(nameof(index));
```

Comment:
Did you discover this app-compat violation with a test or code inspection?  If the latter, can you add a new test?

Reply:
An existing test. Yay tests :)